tools/QuietBackgroundProcesses/DevHome.QuietBackgroundProcesses.PerformanceRecorderEngine/PerformanceRecorderEngine.cpp

```diff
@@ -0,0 +1,669 @@
+    }
+    return std::wstring { packageFullName };
+}
+
+std::optional<std::wstring> TryGetAppUserModelIdFromTokenHelper(HANDLE token)
+{
+    wchar_t aumid[APPLICATION_USER_MODEL_ID_MAX_LENGTH]{};
+    uint32_t aumidLength = ARRAYSIZE(aumid);
+    if (GetApplicationUserModelIdFromToken(token, &aumidLength, aumid) != ERROR_SUCCESS)
+    {
+        return std::nullopt;
+    }
+    return std::wstring { aumid };
+}
+
+std::optional<std::wstring> TryGetProcessName(HANDLE processHandle)
+{
+    static wchar_t s_buffer[MAX_PATH * 2];
+    if (GetModuleFileNameExW(processHandle, nullptr, s_buffer, _countof(s_buffer)) > 0)
+    {
+        return s_buffer;
+    }
+    return std::nullopt;
+}
+
+ProcessPerformanceInfo MakeProcessPerformanceInfo(DWORD processId)
+{
+    auto process = wil::unique_process_handle{ OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION, FALSE, processId) };
+    if (!process)
+    {
+        // We can't open csrss.exe, so we'll just skip processes we can't open
+        return {};
+    }
+
+    auto processPathString = TryGetProcessName(process.get());
+
+    auto path = std::filesystem::path(processPathString.value_or(L""));
+
+    FILETIME createTime, exitTime, kernelTime, userTime;
+    THROW_IF_WIN32_BOOL_FALSE(GetProcessTimes(process.get(), &createTime, &exitTime, &kernelTime, &userTime));
+
+    std::optional<std::wstring> packageFullName;
+    std::optional<std::wstring> aumid;
+    wil::unique_handle processToken;
+    if (OpenProcessToken(process.get(), TOKEN_QUERY, &processToken))
+    {
+        packageFullName = TryGetPackageFullNameFromTokenHelper(processToken.get());
+        aumid = TryGetAppUserModelIdFromTokenHelper(processToken.get());
+    }
+
+    auto info = ProcessPerformanceInfo{};
+    info.process = std::move(process);
+    info.pid = processId;
+    info.name = path.filename().wstring();
+    info.path = path.parent_path().wstring();
+    info.packageFullName = packageFullName;
+    info.aumid = aumid;
+    info.category = GetCategory(info.pid, info.name);
+    info.createTime = createTime;
+
+    // Start times
+    info.startUserTime = userTime;
+    info.startKernelTime = kernelTime;
+
+    info.previousUserTime = userTime;
+    info.currentUserTime = userTime;
+    info.previousKernelTime = kernelTime;
+    info.currentKernelTime = kernelTime;
+
+    return info;
+}
+
+bool UpdateProcessPerformanceInfo(ProcessPerformanceInfo& info)
+{
+    FILETIME createTime, exitTime, kernelTime, userTime;
+    THROW_IF_WIN32_BOOL_FALSE(GetProcessTimes(info.process.get(), &createTime, &exitTime, &kernelTime, &userTime));
+
+    if (exitTime.dwHighDateTime != 0 || exitTime.dwLowDateTime != 0)
+    {
+        info.exitTime = info.exitTime;
+        return false;
+    }
+
+    info.previousUserTime = info.currentUserTime;
+    info.currentUserTime = userTime;
+    info.previousKernelTime = info.currentKernelTime;
+    info.currentKernelTime = kernelTime;
+    return true;
+}
+
+struct cancellation_mechanism
```

Comment:
Why `_`'s in the names?

Reply:
colloquial signal used in std and wil to mean a type is a general mechanism, and not a domain specific type.